feat: Add CEL expression evaluator command

chmouel opened the pull request:
# Changes

This commit introduces a new command `tkn pac cel` that allows users to interactively evaluate CEL (Common Expression Language) expressions. The command is designed to help users test and debug CEL expressions, which are commonly used in Pipelines-as-Code.


https://github.com/user-attachments/assets/e5f7cf83-cf8b-4e50-acbb-9012b1fa5c81


Key features include:

- Interactive and non-interactive modes.
- Support for webhook payloads and headers.
- Provider auto-detection (GitHub, GitLab, Bitbucket Cloud, Bitbucket Data Center, Gitea).
- Direct access to variables as per PAC documentation.
- Cross-platform history with readline experience.
- Comprehensive help and example expressions.

JIRA: https://issues.redhat.com/browse/SRVKP-8026

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Ensure your commit message is clear and informative. Refer to the How to write a git commit message guide. Include the commit message in the PR body rather than linking to an external site (e.g., Jira ticket).

- [ ] ♽ Run make test lint before submitting a PR to avoid unnecessary CI processing. Consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the repository root for an efficient workflow.

- [ ] ✨ We use linters to maintain clean and consistent code. Run make lint before submitting a PR. Some linters offer a --fix mode, executable with make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) are installed).

- [ ] 📖 Document any user-facing features or changes in behavior.

- [ ] 🧪 While 100% coverage isn't required, we encourage unit tests for code changes where possible.

- [ ] 🎁 If feasible, add an end-to-end test. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for details.

- [ ] 🔎 Address any CI test flakiness before merging, or provide a valid reason to bypass it (e.g., token rate limitations).

- If adding a provider feature, fill in the following details:

  - [ ] GitHub App
  - [ ] GitHub Webhook
  - [ ] Gitea/Forgejo
  - [ ] GitLab
  - [ ] Bitbucket Cloud
  - [ ] Bitbucket Data Center

  (update the provider documentation accordingly)

chmouel commented:
/jira

chmouel commented:
## 🤖 AI Review provided by gemini:gemini-2.5-pro: 
Excellent! This is a substantial and well-thought-out feature commit. It adds a powerful new `tkn pac cel` command with great features like provider auto-detection and an interactive shell. The test coverage is also impressive.

Here is a careful analysis of the changes, highlighting potential bugs, spelling mistakes, and code improvements.

### Summary of Analysis

Overall, this is a high-quality commit. The core logic is solid, and the feature is well-implemented with good test coverage. The identified issues are relatively minor and easy to address.

*   **Bugs:** A couple of minor bugs were found, mainly related to redundant code and case-sensitive handling of HTTP headers.
*   **Spelling:** A few minor typos in comments and test data.
*   **Improvements:** Suggestions to make the code more robust and idiomatic.

---

### 🐛 Bugs & Potential Issues

These are items that could cause incorrect behavior under certain conditions.

1.  **Redundant JSON Unmarshaling**

    In the functions that parse provider payloads, `json.Unmarshal` is called on the payload `body` *after* the provider's SDK has already parsed it. This is unnecessary and can be removed for cleaner code.

    *   **File**: `pkg/cmd/tknpac/cel/cel.go`
    *   **In `eventFromGitHub()`**: The line `_ = json.Unmarshal(body, &ghEvent)` is redundant because `github.ParseWebHook` already unmarshals the data.
    *   **In `eventFromGitLab()`**: The block `if err := json.Unmarshal(body, &eventInt); err != nil { ... }` is redundant because `gitlab.ParseWebhook` already handles this.

    **Recommendation:** Remove these redundant `json.Unmarshal` calls. The preceding `ParseWebhook` functions have already done the work.

2.  **Case-Sensitive GitHub Header Lookup**

    HTTP headers are defined as case-insensitive. While most of the new code handles this correctly (e.g., using `strings.EqualFold`), the GitHub event parsing does not.

    *   **File**: `pkg/cmd/tknpac/cel/cel.go`
    *   **In `eventFromGitHub()`**: The line `event.EventType = headers["X-GitHub-Event"]` performs a case-sensitive lookup. If a user provides headers with `x-github-event`, this will fail to find the event type.

    **Recommendation:** Make the lookup case-insensitive to be more robust. You can check for both casings, as is done in the `detectProvider` function:
    ```go
    // In eventFromGitHub()
    event.EventType = headers["X-GitHub-Event"]
    if event.EventType == "" {
        event.EventType = headers["x-github-event"]
    }
    ```

---

### 📝 Spelling and Documentation

These are minor typos and clarifications in comments and documentation.

1.  **Typo in Test Error String**

    There's a casing typo in an expected error message within the tests.

    *   **File**: `pkg/cmd/tknpac/cel/cel_test.go`
    *   **In `TestCommandExecution` and `TestCommandFileHandling`**: The expectation `wantErrContains: "unknown X-Github-Event"` should be `wantErrContains: "unknown X-GitHub-Event"`.

2.  **Misleading Comment**

    A comment in the main command logic is slightly misleading.

    *   **File**: `pkg/cmd/tknpac/cel/cel.go`, in `Command()` function.
    *   **Comment**: `// Create files data structure (empty for now since we don't have a provider)`
    *   **Issue**: At this point in the code, the provider *has* been determined. The reason the `files` data is empty is because file change detection is not implemented in the CLI.
    *   **Recommendation**: Change the comment to be more accurate, for example: `// Create files data structure (always empty in CLI mode)`.

3.  **Minor Typo in Documentation**

    *   **File**: `docs/content/docs/guide/cli.md`
    *   **Text**: "...content of the webhook requests..."
    *   **Recommendation**: Change to "...content of the webhook request...".

---

### 💡 Code Improvements & Idiomatic Go

These are suggestions to make the code cleaner, more robust, or more idiomatic.

1.  **Refactor Bitbucket Provider Detection Logic**

    The logic to distinguish Bitbucket Cloud from Data Center in `detectProvider` works, but it contains a slightly strange-looking repeated `if` block. It can be made more concise.

    *   **File**: `pkg/cmd/tknpac/cel/cel.go`
    *   **Original Code**:
        ```go
        // ...
        if actor, ok := payload["actor"].(map[string]any); ok {
            if _, hasAccountID := actor["account_id"]; hasAccountID {
                return "bitbucket-cloud", nil
            }
        }
        // Bitbucket Data Center has different structure
        if actor, ok := payload["actor"].(map[string]any); ok {
            if _, hasID := actor["id"]; hasID {
                return "bitbucket-datacenter", nil
            }
        }
        // ...
        ```
    *   **Recommendation**: Combine these into a single `if` block for better readability.
        ```go
        if actor, ok := payload["actor"].(map[string]any); ok {
            if _, hasAccountID := actor["account_id"]; hasAccountID {
                return "bitbucket-cloud", nil
            }
            // Heuristic: if it has an `id` but not an `account_id`, assume it's Data Center.
            if _, hasID := actor["id"]; hasID {
                return "bitbucket-datacenter", nil
            }
        }
        ```

2.  **Explicitly State `files.*` Limitation in Help Template**

    The help template is very good, but the note about `files.*` functionality could be slightly punchier.

    *   **File**: `pkg/cmd/tknpac/cel/templates/help.tmpl`
    *   **Suggestion**: The current text is clear but a bit long. Consider combining the two lines about file variables.
        ```diff
        -  📁 Files (files.* - ❌ always empty in CLI):
        -    • files.all, files.added, files.modified, etc.
        -    ⚠️  Note: fileChanged, fileDeleted, fileModified and all files.* variables are not implemented in the CLI and will always be empty.
        +  📁 Files (files.*):
        +    ⚠️  Note: File-related variables (e.g., `files.all`) and functions (e.g., `fileChanged`) are not supported in the CLI and will be empty or false.
        ```
    This is a stylistic choice; the original is also perfectly fine.

### Final Verdict

This is an excellent commit that adds a very useful feature. The implementation is robust, especially with the fallback from `readline` to `survey` and the thoughtful provider auto-detection logic. The identified issues are minor and do not detract from the overall quality of the contribution. Congratulations to the author on a job well done

chmouel commented:
/retest

zakisk commented:
/lgtm

zakisk commented:
/retest

zakisk commented:
it is LGTM, do I merge or you wanna make more changes?

chmouel commented:
@zakisk i think wait for it, there is some edge case that makes it buggy compared to live event... i'll let you know if we keep it "as is"

chmouel commented:
/gemini review

zakisk commented on docs/content/docs/guide/cli.md:
```diff
@@ -409,6 +409,111 @@ You can specify a different directory than the current one by using the -d/--dir
+
+The payload is the JSON content of the webhook request, The headers file supports multiple formats:
+
+1. **Plain HTTP headers format** (as shown above)
```
you said "(as shown above), but where?

zakisk replied:
```suggestion
1. **Plain HTTP headers format** (as shown below)
```

may be you would add example headers here

zakisk commented:
this looks good to me but I am holding merge if you wanna address Gemini reviews.

chmouel commented:
/gemini review

zakisk commented:
/lgtm

zakisk commented on pkg/cmd/tknpac/cel/cel.go:
```diff
@@ -0,0 +1,1099 @@
+						"deleted":  []string{},
+						"modified": []string{},
+						"renamed":  []string{},
+					}
```
@chmouel it makes I guess? or feel free to merge!

chmouel commented:
/retest

chmouel commented:
@zakisk would you like to do a final review, since i have modified it quite a bit with regard to the readline completions

zakisk reviewed (approved):
Looks good!

zakisk commented:
@chmouel how to exit CEL evaluator? I needed to Ctrl+C which is not ideal I think

chmouel commented:
Ctrl+d is the usually what is used to exit in a CLI but i can add a /exit

chmouel committed:
feat: Add CEL expression evaluator command

This commit introduces a new command `tkn pac cel` that allows
users to interactively evaluate CEL (Common Expression Language)
expressions. The command is designed to help users test and debug
CEL expressions, which are commonly used in Pipelines-as-Code.

Key features include:

- Interactive and non-interactive modes.
- Support for webhook payloads and headers.
- Provider auto-detection (GitHub, GitLab, Bitbucket Cloud,
  Bitbucket Data Center, Gitea).
- Direct access to variables as per PAC documentation.
- Cross-platform history with readline experience.
- Comprehensive help and example expressions.

Note: this evaluator is primarily intended for repository or GitHub app
admins, as it requires access to webhook payloads and headers.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
added one in last push 
<img width="1358" height="622" alt="image" src="https://github.com/user-attachments/assets/fcf08ceb-f86b-4dd1-a2d9-d6ec09a8234f" />
